subsys: power: device implicit depends on CONFIG_SYS_POWER_MANAGEMENT

**Describe the bug**
When try using subsys:shell on a device that have Power Management capabilities **device_pm_state_str** doesn't compile. Some investigation shows that subsys/power/device.c compile only when CONFIG_SYS_POWER_MANAGEMENT is enabled.
However, SYS_POWER_MANAGEMENT and DEVICE_POWER_MANAGEMENT are independent on the subsys/power/Kconfig.

https://github.com/zephyrproject-rtos/zephyr/blob/016b5422d90e9d3405eda22da4fda23d6a13a415/subsys/shell/modules/device_service.c#L101-L110

https://github.com/zephyrproject-rtos/zephyr/blob/016b5422d90e9d3405eda22da4fda23d6a13a415/subsys/power/device.c#L13

**To Reproduce**
See #28094

**Expected behavior**
If sys PM and dev PM are independent things, the subsys/power/device.c should be adjusted. Otherwise, when users select DEVICE_POWER_MANAGEMENT, the SYS_POWER_MANAGEMENT should be selected to avoid problems. However, it is expected that some SoC does't have proper clocks and PM implemented on Zephyr and this should be treat accordly.

The path to add a new SoC in general is a long journey and maybe flag like SOC_SERIES_HAS_PM could help to avoid problems. A gracious build error message with some information that device driver PM could not be enabled because SoC doesn't have PM implemented will be enough to uses understand the situation.

**Impact**
Device Power Management is not independent right now on Zephy. Any Power Management feature will only work for those devices that define **core_devices** hard coded at `device.c` and there is no possible to use device PM on boards that don't implement sys PM.

https://github.com/zephyrproject-rtos/zephyr/blob/016b5422d90e9d3405eda22da4fda23d6a13a415/subsys/power/device.c#L18-L54

This could bring a lot of unnecessary issues after release 2.40.

Ex.: Try use a SPI_FLASH_AT45 NOR with PM activated on a SoC that doesn't have sys PM implemented.


**Logs and console output**
```
home/gfbudke/zephyros/mainline/zephyr/subsys/power/device.c
/home/gfbudke/zephyros/mainline/zephyr/subsys/power/device.c:53:2: error: #error "Add SoC's core devices list for PM"
   53 | #error "Add SoC's core devices list for PM"
      |  ^~~~~
/home/gfbudke/zephyros/mainline/zephyr/subsys/power/device.c:68:32: error: 'MAX_PM_DEVICES' undeclared here (not in a function)
   68 | static device_idx_t pm_devices[MAX_PM_DEVICES];
      |                                ^~~~~~~~~~~~~~
In file included from ../include/sched_priq.h:9,
                 from ../include/kernel_includes.h:23,
                 from ../include/kernel.h:17,
                 from ../include/zephyr.h:18,
                 from /home/gfbudke/zephyros/mainline/zephyr/subsys/power/device.c:7:
/home/gfbudke/zephyros/mainline/zephyr/subsys/power/device.c: In function 'sys_pm_create_device_list':
/home/gfbudke/zephyros/mainline/zephyr/subsys/power/device.c:163:22: error: 'core_devices' undeclared (first use in this function); did you mean 'pm_devices'?
  163 |  num_pm = ARRAY_SIZE(core_devices);
      |                      ^~~~~~~~~~~~
../include/sys/util.h:67:66: note: in definition of macro 'ZERO_OR_COMPILE_ERROR'
   67 | #define ZERO_OR_COMPILE_ERROR(cond) ((int) sizeof(char[1 - 2 * !(cond)]) - 1)
      |                                                                  ^~~~
../include/sys/util.h:98:11: note: in expansion of macro 'IS_ARRAY'
   98 |  ((long) (IS_ARRAY(array) + (sizeof(array) / sizeof((array)[0]))))
      |           ^~~~~~~~
/home/gfbudke/zephyros/mainline/zephyr/subsys/power/device.c:163:11: note: in expansion of macro 'ARRAY_SIZE'
  163 |  num_pm = ARRAY_SIZE(core_devices);
      |           ^~~~~~~~~~
/home/gfbudke/zephyros/mainline/zephyr/subsys/power/device.c:163:22: note: each undeclared identifier is reported only once for each function it appears in
  163 |  num_pm = ARRAY_SIZE(core_devices);
      |                      ^~~~~~~~~~~~
../include/sys/util.h:67:66: note: in definition of macro 'ZERO_OR_COMPILE_ERROR'
   67 | #define ZERO_OR_COMPILE_ERROR(cond) ((int) sizeof(char[1 - 2 * !(cond)]) - 1)
      |                                                                  ^~~~
../include/sys/util.h:98:11: note: in expansion of macro 'IS_ARRAY'
   98 |  ((long) (IS_ARRAY(array) + (sizeof(array) / sizeof((array)[0]))))
      |           ^~~~~~~~
/home/gfbudke/zephyros/mainline/zephyr/subsys/power/device.c:163:11: note: in expansion of macro 'ARRAY_SIZE'
  163 |  num_pm = ARRAY_SIZE(core_devices);
      |           ^~~~~~~~~~
At top level:
/home/gfbudke/zephyros/mainline/zephyr/subsys/power/device.c:68:21: warning: 'pm_devices' defined but not used [-Wunused-variable]
   68 | static device_idx_t pm_devices[MAX_PM_DEVICES];
      |                     ^~~~~~~~~~
[58/148] Building C object zephyr/arch/common/CMakeFiles/isr_tables.dir/isr_tables.c.obj
```
**Environment (please complete the following information):**
 - OS: Linux
 - Zephyr SDK 0.11.3
 - 2.4.0-rc1

--- a/subsys/power/Kconfig
+++ b/subsys/power/Kconfig
@@ -92,5 +92,9 @@ config DEVICE_POWER_MANAGEMENT
 	  like turning off device clocks and peripherals. The device drivers
 	  may also save and restore states in these hook functions.
 
+config PM_MAX_DEVICES
+	int "Max number of devices support power management"
+	depends on DEVICE_POWER_MANAGEMENT
+	default 15
 
 endmenu

--- a/subsys/power/device.c
+++ b/subsys/power/device.c
@@ -21,37 +21,24 @@ LOG_MODULE_DECLARE(power);
  * to build the device list based on devices power
  * and clock domain dependencies.
  */
+
+__weak const char *const z_pm_core_devices[] = {
 #if defined(CONFIG_SOC_FAMILY_NRF)
-#define MAX_PM_DEVICES	15
-static const char *const core_devices[] = {
 	"CLOCK",
 	"sys_clock",
 	"UART_0",
-};
 #elif defined(CONFIG_SOC_SERIES_CC13X2_CC26X2)
-#define MAX_PM_DEVICES	15
-static const char *const core_devices[] = {
 	"sys_clock",
 	"UART_0",
-};
 #elif defined(CONFIG_SOC_SERIES_KINETIS_K6X)
-#define MAX_PM_DEVICES		1
-static const char *const core_devices[] = {
 	DT_LABEL(DT_INST(0, nxp_kinetis_ethernet)),
-};
 #elif defined(CONFIG_NET_TEST)
-#define MAX_PM_DEVICES		1
-static const char *const core_devices[] = {
 	"",
-};
 #elif defined(CONFIG_SOC_SERIES_STM32L4X) || defined(CONFIG_SOC_SERIES_STM32WBX)
-#define MAX_PM_DEVICES	1
-static const char *const core_devices[] = {
 	"sys_clock",
-};
-#else
-#error "Add SoC's core devices list for PM"
 #endif
+	NULL
+};
 
 /* Ordinal of sufficient size to index available devices. */
 typedef uint16_t device_idx_t;
@@ -65,7 +52,7 @@ static const struct device *all_devices;
 /* Indexes into all_devices for devices that support pm,
  * in dependency order (later may depend on earlier).
  */
-static device_idx_t pm_devices[MAX_PM_DEVICES];
+static device_idx_t pm_devices[CONFIG_PM_MAX_DEVICES];
 
 /* Number of devices that support pm */
 static device_idx_t num_pm;
@@ -149,7 +136,7 @@ void sys_pm_resume_devices(void)
 void sys_pm_create_device_list(void)
 {
 	size_t count = z_device_get_all_static(&all_devices);
-	device_idx_t pmi;
+	device_idx_t pmi, core_dev;
 
 	/*
 	 * Create an ordered list of devices that will be suspended.
@@ -160,9 +147,15 @@ void sys_pm_create_device_list(void)
 	__ASSERT_NO_MSG(count <= DEVICE_IDX_MAX);
 
 	/* Reserve initial slots for core devices. */
-	num_pm = ARRAY_SIZE(core_devices);
+	core_dev = 0;
+	while (z_pm_core_devices[core_dev]) {
+		core_dev++;
+	}
 
-	for (pmi = 0; (pmi < count) && (num_pm < MAX_PM_DEVICES); pmi++) {
+	num_pm = core_dev;
+	__ASSERT_NO_MSG(num_pm <= CONFIG_PM_MAX_DEVICES);
+
+	for (pmi = 0; pmi < count; pmi++) {
 		device_idx_t cdi = 0;
 		const struct device *dev = &all_devices[pmi];
 
@@ -174,8 +167,8 @@ void sys_pm_create_device_list(void)
 		/* Check if the device is a core device, which has a
 		 * reserved slot.
 		 */
-		while (cdi < ARRAY_SIZE(core_devices)) {
-			if (strcmp(dev->name, core_devices[cdi]) == 0) {
+		while (z_pm_core_devices[cdi]) {
+			if (strcmp(dev->name, z_pm_core_devices[cdi]) == 0) {
 				pm_devices[cdi] = pmi;
 				break;
 			}
@@ -183,7 +176,7 @@ void sys_pm_create_device_list(void)
 		}
 
 		/* Append the device if it doesn't have a reserved slot. */
-		if (cdi == ARRAY_SIZE(core_devices)) {
+		if (cdi == core_dev) {
 			pm_devices[num_pm++] = pmi;
 		}
 	}

